[FEAT/#101] Group / Group 화면 최소기능 완성

TaewoongR opened the pull request:
- closed #101 

## *📍 Work Description*

- Group 화면 최소기능 완성

## *📸 Screenshot*

https://github.com/user-attachments/assets/395b6057-2af5-4b29-ac46-0780762b3d95

https://github.com/user-attachments/assets/e6cb5305-c66b-4d33-ba18-590ee1cb2b43

<img src="https://github.com/user-attachments/assets/90674ad9-7c82-49a7-be34-820b7d3f7022" width=300>

https://github.com/user-attachments/assets/2d8347bd-6662-4bcf-9412-1f26d843bab6

<img src="https://github.com/user-attachments/assets/24c5283e-2ec6-485a-a591-01f0812410ad" width=300>

<img src="https://github.com/user-attachments/assets/b6304e5b-b37d-4e09-af8f-777bac5c7726" width=300>


## *📢 To Reviewers*

### 1️⃣ 공통 사항

- **뷰모델과 구조**
  - 각 화면에 대응되는 ViewModel은 `state`, `intent`, `sideEffect`를 관리.
  - 화면 구성은 `Screen → Content → Field`로 나뉨.
    - **Screen**: `intent`와 `sideEffect`를 처리하며, `Content`에 `state`를 전달.
    - **Content, Field**: 전달받은 `state`로 UI를 구성.
  - `Scaffold`의 시작점은 `Content`로 설정.

- **SideEffect 수집**
  - **Screen**에서 `sideEffect`를 수집하며, 화면의 생명주기 동안 유지.
  - `collectAsStateWithLifecycle`과 유사한 커스텀 함수를 제작하여 사용.
    - `Channel`로 초기화된 `sideEffect`를 `Flow`로 변환.
    - 초기값 없는 상태를 보완하기 위해 `remember`와 `collectAsState`를 합성한 상태 저장 함수 제작.
    - 초기값은 아무 동작도 하지 않는 `Idle`로 설정.

- **공통 사용 리소스**
  - 모든 ViewModel은 도메인의 `GroupRepository`에서 정의된 함수 사용.
  - `usecase` 미적용.
  - **UserPreferenceDataStore 활용** (단, `GroupCreationScreen`은 제외).
    - **GroupScreen**: User ID로 속한 그룹 조회.
    - **GroupJoinScreen**: 그룹 참여 시, User ID를 Firestore의 `group/members`와 `user/group` 필드에 저장.
    - **GroupDetailScreen**: 그룹 편집 버튼 노출 조건으로 `adminUser` ID와 User ID 비교.
    - **GroupCreationScreen**: 그룹 생성 시 `adminUser` ID를 User ID로 설정.

---

### 2️⃣ GroupScreen

- **기능**
  - User ID 기반으로 사용자가 속한 그룹만 화면에 표시.
  - 내부 화면 이동 후 복귀 시 그룹 리스트를 다시 로드하여 최신 데이터 반영.
  - 특정 그룹을 터치하면 `groupId`를 `Navigation`으로 전달하여 **GroupDetailScreen**으로 이동.

---

### 3️⃣ GroupJoinScreen

- **초대 코드 입력**
  - Firestore의 `inviteCodes` 컬렉션에서 `document(id)`를 탐색.
  - 존재할 경우, 해당 문서의 `group` 필드에서 `groupId`를 가져옴.
  - 이후, `groupId`를 사용하여 그룹 객체를 탐색.

- **참여 버튼 동작**
  - Firestore에서 `group/members`에 User ID를 추가.
  - 동시에 `user/group` 필드에 `groupId` 추가.

- **오류 처리**
  - **그룹 찾기 실패**: 버튼 아래에 “그룹이 존재하지 않음” 문구 표시.
  - **인터넷 문제**: Toast 메시지로 오류 알림.

---

### 4️⃣ GroupCreationScreen

- **입력 필드**
  - 사진 선택, 제목, 내용 입력 필드가 존재하며 모두 필수 입력.
    - **제목**: 2~20자 제한.
    - **내용**: 최소 10자 이상 입력 요구.

- **생성 요청 조건**
  - 그룹 객체의 프로퍼티 설정:
    - `adminUser`: User ID.
    - `groupId`: 랜덤 UUID.
    - `members`: User ID 포함.
    - `createdAt`: `java.util.Date()` 값.

- **Firestore 연동**
  - 그룹 객체를 `group` 컬렉션에 추가 요청.
    - **성공 시**: Toast 메시지 표시 후 뒤로가기 전환.
    - **실패 시**: Toast 메시지로 오류 알림.

---

### 5️⃣ GroupDetailScreen

- **그룹 정보 가져오기**
  - **GroupScreen**에서 전달받은 `groupId`를 사용하여 Firestore에서 `group` 객체를 요청.
  - 요청한 `group` 객체의 `members`(String 리스트)를 통해 각 User 객체를 Firestore에서 추가로 요청.
  - 가져온 `group` 객체와 `user` 객체를 화면에 표시.

- **초대 코드 발급**
  - **초대 코드 복사**: 초대코드를 클립보드에 복사한 뒤 Toast 메시지로 알림.
  - 초대 코드 생성 과정:
    1. 해당 `groupId`에 이미 발급된 `inviteCodes` 컬렉션 탐색.
       - 존재한다면 해당 컬렉션의 `document(id)` 반환.
    2. 존재하지 않을 경우:
       - 새 `inviteCodes` 컬렉션 생성.
       - `document(id)`는 랜덤 UUID로 생성.
       - `field` 구성:
         - `expiredBy`: 오늘로부터 일주일 뒤의 날짜(Firebase의 `Timestamp` 타입 사용).
         - `group`: 해당 `groupId`.

- **에피소드 데이터 가져오기**
  - **User** 객체 요청 이후, `groupId`를 통해 관련된 에피소드 데이터를 추가로 요청.
  - **EpisodeRepository**를 통해 Firestore에서 에피소드 객체를 가져옴.
  - 가져온 에피소드 데이터를 화면의 탭 오른쪽에 표시.

- **에피소드 정렬**
  - 최초 정렬 방식은 **최신순**.
  - 드롭다운 메뉴를 통해 다음과 같은 방식으로 정렬 가능:
    - 최신순
    - 과거순
    - 날짜순

- **GroupEditScreen으로 이동**
  - **조건**: `adminUser` ID가 현재 유저의 ID와 일치하는 경우에만, 탑바의 액션 버튼에 이동 버튼 표시.
  - 이동 시 `GroupEditScreen`에 `groupId`를 전달.

---

### 6️⃣ GroupEditScreen

- **그룹 데이터 가져오기**
  - 전달받은 `groupId`를 사용하여 Firestore에서 해당 `group` 객체를 요청.
  - 가져온 데이터를 활용하여 다음 필드에 값 부여:
    - 사진 선택 필드
    - 타이틀 `TextField`
    - 내용 `TextField`

- **편집 제약사항**
  - **GroupCreationScreen**의 제약사항과 동일:
    - 제목: 2~20자.
    - 내용: 10자 이상.
    - 사진 선택 필수.

- **그룹 정보 편집**
  - 편집 버튼 클릭 시, 기존 `group` 객체를 복사하여 수정된 사항 반영.
  - Firestore에 업데이트 요청:
    - 객체에 저장된 `groupId`를 통해 `group` 컬렉션에서 해당 그룹 추적.
    - 덮어쓰기 방식으로 업데이트 수행.

- **결과 처리**
  - **성공 시**: Toast 메시지 표시 및 이전 화면으로 전환.
  - **실패 시**: Toast 메시지로 오류 알림.

## ⏲️Time

    - 24시간

TaewoongR committed:
feat/#101 : 그룹 참여 함수 repository 에 선언

- firestore : group collection에 멤버 추가
- firestore : user collection에 그룹 추가

TaewoongR committed:
feat/#101 : 유저 아이디 뷰모델로 가져오기

- gradle 의존성 core:datastore 추가
- 뷰모델에서 아이디 옵저빙 함수 선언

TaewoongR committed:
feat/#101 : 그룹 아이디 코드로 그룹 정보 get 함수 선언

- 그룹이 존재하지 않을 시 exception throw

TaewoongR committed:
feat/#101 : 그룹 참여 서버와 연결 진행

- 참여 화면을 위한 intent, state, sideEffect 정의
- 현재 firestore의 그룹 아이디 자체를 입력해야 그룹 정보를 가져올 수 있음
- 코드 입력 후 그룹 정보 출력 가능
- 참여하기 미완성

TaewoongR committed:
refactor/#101 : 그룹코드 대신 초대코드를 통해 그룹 정보 가져오기

TaewoongR committed:
feat/#101 : 그룹 참여 + 참여 성공 후 뒤로가기 기능 추가

- firestore의 user, group 컬렉션의 필드에 각각 참여 그룹과 멤버의 DocumentReference를 붙힘

TaewoongR committed:
feat/#101 : type-safe nav basic type 전달

- 그룹 상세와 그룹 수정 화면에 firestore의 group Id 전달

TaewoongR committed:
fix/#101 : custom Card 컴포넌트 파일 위치 변경

- feature:mygroup -> core:designsystem
- designsystem 에 coil 의존성 추가

TaewoongR committed:
feat/#101 : 그룹 상세 화면 뷰모델, intent, state 선언

TaewoongR committed:
fix/#101 : develop 브랜치 rebase conflict 작업

TaewoongR committed:
fix/#101 : intent, state, sideEffect 파일 디렉토리 분류

TaewoongR committed:
feat/#101 : 그룹 아이디로 그룹 정보 가져오기

- data:mygroup 에서 그룹 아이디로 그룹 정보 가져오는 함수 생성
- repository를 통해 viewmodel에서 사용

TaewoongR committed:
fix/#101 : 그룹 메인화면, 그룹 참여화면 mvi 수정

- sideEffect의 책임을 intent가 겸하고 있던 부분을 수정하여 각 책임을 분리
- intent는 유저가 입력하는 이벤트만을 처리
- sideEffect는 이벤트에 의한 작동으로 분리

TaewoongR committed:
fix/#101 : 토스트 직후 화면전환 사이에 딜레이 적용

- 딜레이 없이는 Screen에서 이전 sideEffect 수집 전에 다음 sideEffect를 수집 후 수행

TaewoongR committed:
feat/#101 : 화면 전환 후 그룹메인 화면으로 복귀시 최신 그룹 리스트 띄우기

- state의 isInitializing을 딜레이를 걸어 약간 뒤에 true로 변환 : screen 컴포저블에서 state 수집하여 즉시 그룹을 로드할 수 있기에 해당 컴포저블을 벗어난 뒤에 값 변환

TaewoongR committed:
feat/#101 : GroupJoinState 프로퍼티 이름 변경

- 이름에서 Join 제외

TaewoongR committed:
feat/#101 : GroupDetailSideEffect 확장

TaewoongR committed:
feat/#101 : 화면 흐름 설정

- 1. 디테일 화면이 최초에 그룹 아이디를 전달받음
- 2. 그룹 아이디를 뷰모델에 캐싱
- 3. 그룹 아이디를 통해 그룹 객체 로딩
- 4. 편집 화면으로 이동시 되돌아올 때 다시 로딩하도록 설정

TaewoongR committed:
fix/#101 : GroupInfoCard modifier 수정

- 내부에서 고정 modifier가 아닌 외부에서 modifier 설정

TaewoongR committed:
fix/#101 : 그룹정보 카드 부연설명 간격 수정

TaewoongR committed:
fix/#101 : 그룹 설명 박스 배경 색상 변경

TaewoongR committed:
feat/#101 : 참여 코드 발급 기능 추가

- firestore 에 참여 코드 발급 요청 : 기존에 있으면 기존것으로 반환, 없으면 새로 생성 후 반환
- 참여코드는 클립보드에 복사되며 토스트로 이를 알림

TaewoongR committed:
feat/#101 : 그룹 참여 멤버 리스트 적용

- 그룹 데이터 클래스에서 members 에 저장된 DocumentReference 문자열을 각각 사용하여 firestore에 유저 정보 객체를 요청하고 그 객체를 받아 화면에 띄움

TaewoongR committed:
feat/#101 : 그룹 탈퇴 기능 추가

- 1. 그룹 탈퇴 버튼을 누르면 다이얼로그를 먼저 띄운다.
- 2. 취소, 뒤로가기를 누르면 다이얼로그가 사라지고 아무런 동작을 안한다.
- 3. 확인시 firestore에게 user에 저장된 그룹과 group에 저장된 멤버에서 데이터를 삭제한다.
- 4. 삭제 후 다이얼로그를 닫고 그룹 메인 화면으로 이동한다.

TaewoongR committed:
feat/#101 : 그룹의 에피소드 리스트 화면 선언

- 1. 그룹 설명 화면 정보가 전부 로드가 된 후에 그룹 아이디를 통해 그룹의 에피소드들을 firestore에 요청하여 가져온다.
- 2. 에피소드 정렬 메뉴의 선택 목록에 따라 에피소드의 정렬이 달라진다. 초기값은 최신순으로 정렬되어 있다.

TaewoongR committed:
feat/#101 : 유저가 그룹의 어드민 유저이면 편집 아이콘 활성화, 그렇지 않다면 비활성화

TaewoongR committed:
fix/#101 : 임시 로직 삭제 예정 주석 추가

TaewoongR committed:
fix/#101 : 불필요한 constructor 제거 후 class -> object 변환

TaewoongR committed:
feat/#101 : intent, state, sideEffect, viewmodel 정의

- 1. 그룹상세 화면에서 그룹 아이디를 받아온다.
- 2. 그룹 아이디를 통해 그룹 객체를 firestore에서 받아온다.
- 3. 모든 입력을 기입한 후 버튼을 누르면 수정된 그룹 정보 업데이트를 시도한다.
- 4. 업데이트에 성공하면 뒤로가며, 실패시 토스트 메세지 출력

TaewoongR committed:
feat/#101 : 그룹 편집 화면 기능 적용

- 1. 편집화면으로 이동하면 그룹 이미지, 제목, 내용이 뜹니다.
- 2. 제목과 내용을 수정하고 버튼을 누르면 수정된 내용을 바탕으로 firestore에 업데이트(덮어씌우기)를 요청합니다.
- 3. 성공하면 이전 화면으로, 실패시 토스트 메세지를 띄웁니다.

TaewoongR committed:
feat/#101 : 그룹 생성 화면 완성

TaewoongR committed:
fix/#101 : for ktlint

TaewoongR committed:
feat/#101 : firestore createGroup 함수 수정

- 그룹 생성 시 firestore의 user 컬렉션에 groupId 데이터와 group 컬렉션에 userId 데이터 저장

TaewoongR committed:
fix/#101 : for ktlint

TaewoongR committed:
fix/#101 : 그룹화면 유저 아이디 정보 접근 수정

- 임의로 하드코딩된 아이디 값을 가져오는 방식 -> datastore에서 유저아이디 가져오기

TaewoongR committed:
fix/#101 : 편집, 생성화면 입력 정보 제한

- 제목 : 2~24글자
- 내용 : 10글자 이상
- 사진 필수

TaewoongR committed:
fix/#101 : for ktlint

TaewoongR committed:
fix/#101 : 초대 코드 발급시 유효기간 7일로 설정

Ameri-Kano commented:
맨 위에 #101 추가해야 할 것 같아요!

shinythink commented on data/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/GroupRepositoryImpl.kt:
```diff
@@ -32,12 +49,83 @@ class GroupRepositoryImpl @Inject constructor(private val database: FirebaseFire
+		throw e
+	}
+
+	override suspend fun leaveGroup(userId: String, groupId: String) {
+		groupCollection.document(groupId).update(
+			FirestoreConstants.FIELD_MEMBERS,
+			FieldValue.arrayRemove(userCollection.document(userId)),
+		).await()
+
+		userCollection.document(userId).update(
+			FirestoreConstants.FIELD_GROUPS,
+			FieldValue.arrayRemove(groupCollection.document(groupId)),
+		).await()
+	}
```
둘 중 하나의 작업만 성공하는 경우에 대비해야 할 것 같습니다. 트랜잭션으로 묶으면 좋을 것 같습니다.

TaewoongR replied:
그러네요 트랜잭션이 적용되어야 할 함수가 전혀 안되어있는데 이부분 적용하겠습니다.
정말 감사해요!~

shinythink commented on data/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/GroupRepositoryImpl.kt:
```diff
@@ -32,12 +49,83 @@ class GroupRepositoryImpl @Inject constructor(private val database: FirebaseFire
+		throw e
+	}
+
+	override suspend fun joinGroup(userId: String, groupId: String) {
+		groupCollection.document(groupId).update(
+			FirestoreConstants.FIELD_MEMBERS,
+			// 기존 유저 리스트에 덮어씌우지 않고 추가
+			FieldValue.arrayUnion(userCollection.document(userId)),
+		).await()
+
+		userCollection.document(userId).update(
+			FirestoreConstants.FIELD_GROUPS,
+			FieldValue.arrayUnion(groupCollection.document(groupId)),
+		).await()
+	}
```
이것도 트랜젝션으로 묶으면 좋을 것 같습니다.

shinythink commented on feature/login/src/main/java/com/boostcamp/mapisode/login/AuthViewModel.kt:
```diff
@@ -85,7 +85,7 @@ class AuthViewModel @Inject constructor(
 			try {
 				if (uiState.value.nickname.isBlank()) throw IllegalArgumentException("닉네임을 입력해주세요.")
-				if (uiState.value.profileUrl.isBlank()) throw IllegalArgumentException("프로필 사진을 선택해주세요.")
+				// if (uiState.value.profileUrl.isBlank()) throw IllegalArgumentException("프로필 사진을 선택해주세요.")
```
감사합니다 ㅠㅠ

shinythink commented on feature/mygroup/build.gradle.kts:
```diff
@@ -11,4 +11,6 @@ android {
 	implementation(libs.bundles.coil)
 	implementation(projects.domain.mygroup)
+	implementation(projects.domain.episode)
+	implementation(projects.core.datastore)
```
이건 플러그인에 추가되어 있어서 추가하지 않아도 괜찮을 것 같습니다.

TaewoongR replied:
datastore가 있었네요..감사합니다 ㅠㅠ

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupCreationScreen.kt:
```diff
@@ -41,44 +118,135 @@ fun GroupCreationScreen(onBackClick: () -> Unit) {
+	paddingValues: PaddingValues,
+	onGroupEditClick: (title: String, content: String, imageUrl: String) -> Unit,
+) {
+	var name by remember { mutableStateOf("") }
+	var description by remember { mutableStateOf("") }
+	var profileUrl by remember { mutableStateOf("") }
```
이것들은 viewmodel에서 관리하지 않아도 되는 녀석들인가요??

TaewoongR replied:
버튼을 누르지 않는 이상 굳이 뷰모델에서 세가지 데이터를 캐싱할 필요가 있을까 싶어서 컴포저블에서 관리하고 있습니다.

뷰모델에서 관리하게되면 한글자 추가 또는 삭제시마다 업데이트 되어야하는데 오히려 복잡해질 것 같아요.

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupEditScreen.kt:
```diff
@@ -58,29 +127,139 @@ fun GroupEditScreen(
+	group: GroupModel,
+	onGroupEditClick: (title: String, content: String, imageUrl: String) -> Unit,
+) {
+	var name by remember { mutableStateOf(group.name) }
+	var description by remember { mutableStateOf(group.description) }
+	var profileUrl by remember { mutableStateOf(group.imageUrl) }
```
rememberSaveable로 하지 않아도 괜찮나요??

TaewoongR replied:
습관성 remember ... rememberSaveable 반영하겠음다

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/viewmodel/GroupJoinViewModel.kt:
```diff
@@ -0,0 +1,94 @@
+			try {
+				val group = groupRepository.getGroupByInviteCodes(inviteCodes)
+				intent { copy(isGroupExist = true, group = group) }
+				Timber.d("group: $currentState")
```
여기 timber 없애도 괜찮을 것 같아요!

TaewoongR replied:
저기 숨어있었네요. 삭제 쓱삭 진행하겠음다

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/viewmodel/GroupDetailViewModel.kt:
```diff
@@ -0,0 +1,176 @@
+						episodes = episodes,
+					)
+				}
+				Timber.e("episodes: $episodes")
```
이 timber도 없앨 수 있을 것 같아요

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupJoinScreen.kt:
```diff
@@ -180,7 +221,8 @@ fun GroupJoinScreen(
 					MapisodeDivider(direction = Direction.Horizontal, thickness = Thickness.Thin)
 					Spacer(modifier = Modifier.padding(5.dp))
 					MapisodeFilledButton(
-						onClick = {},
+						modifier = Modifier.fillMaxWidth(),
+						onClick = { onJoinGroup() },
 						text = "참여하기",
 						showRipple = true,
 					)
```
이 버튼 좀 더 넓히는 건 어떤가요??

TaewoongR replied:
fillMaxWidth가 최대 너비인데 더 넓힌다는게 어느 부분인거죠?

shinythinking reviewed (approved):
와 노력이 느껴집니다! 
수고 많으셨습니다!

TaewoongR committed:
fix/#101 : 중복 의존성 제거

Ameri-Kano commented on core/firebase/src/main/java/com/boostcamp/mapisode/firebase/firestore/FirestoreConstants.kt:
```diff
@@ -4,6 +4,7 @@ object FirestoreConstants {
 	const val COLLECTION_EPISODE = "episode"
 	const val COLLECTION_GROUP = "group"
 	const val COLLECTION_USER = "user"
+	const val COLLECTION_INVITE_CODES = "inviteCodes"
```
👍

TaewoongR committed:
fix/#101 : 코드 리뷰 피드백 반영

- 중복 의존성 제거
- timber 제거
- firestore 요청 시 Read 제외 한 함수에서 두번 이상의 CRUD 가 존재할 시 transaction 적용

Ameri-Kano commented on core/navigation/src/main/java/com/boostcamp/mapisode/navigation/GroupRoute.kt:
```diff
@@ -7,11 +7,11 @@ sealed interface GroupRoute : Route {
 	@Serializable
-	data object Detail : GroupRoute
+	data class Detail(val groupId: String) : GroupRoute
 
 	@Serializable
 	data object Creation : GroupRoute
 
 	@Serializable
-	data object Edit : GroupRoute
+	data class Edit(val groupId: String) : GroupRoute
```
이렇게 Navigation 클래스에 파라미터를 전달하면 해당 아이디로 리포지토리에 요청해서 데이터를 모델로 받아오게 되나요?

TaewoongR committed:
fix/#101 : GroupCreationScreen 에서 state 사용

- 생성 실패시, 실패 상태를 뷰에 전달하고 상태에 따른 이벤트 실행

Ameri-Kano commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupJoinScreen.kt:
```diff
@@ -44,18 +43,66 @@ import com.boostcamp.mapisode.designsystem.compose.MapisodeTextField
+import com.boostcamp.mapisode.mygroup.sideeffect.rememberFlowWithLifecycle
+import com.boostcamp.mapisode.mygroup.state.GroupJoinState
+import com.boostcamp.mapisode.mygroup.viewmodel.GroupJoinViewModel
 import com.boostcamp.mapisode.mygroup.R as S
```
조금 더 구분되는 이름이었으면 좋겠어요!

Ameri-Kano commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupScreen.kt:
```diff
@@ -26,55 +28,83 @@ import com.boostcamp.mapisode.designsystem.compose.MapisodeIcon
+import com.boostcamp.mapisode.mygroup.sideeffect.rememberFlowWithLifecycle
+import com.boostcamp.mapisode.mygroup.state.GroupState
 import com.boostcamp.mapisode.mygroup.viewmodel.GroupViewModel
 import com.boostcamp.mapisode.mygroup.R as S
```
이것도!

Ameri-Kano reviewed (approved):
너무 고생 많으셨어요!!!

TaewoongR committed:
fix/#101 : for ktlint
